[Fix][MapTool] Circle with M value

lbartoletti opened the pull request:
## Description

Follow up #42812, #42986 and #43060

This PR fixes M support for circle map tool.

I will open two other PR for Ellipse and Quadrilateral.

lbartoletti committed:
Adds setMValueFromPoints in QgsCircle

lbartoletti committed:
Adds and reorganizes tests for QgsMapToolCircle

lbartoletti committed:
reintroduce qgssettingsregistrycore

rouault reviewed (commented):
That's quite a lot of testing code that is more or less a copy&paste of the same file. Couldn't the 4 test files merged into a single one with some parametrization of the test cases to test each of the 4 variants (XY, XYM, XYZ, XYZM) ?

lbartoletti commented:
> That's quite a lot of testing code that is more or less a copy&paste of the same file. Couldn't the 4 test files merged into a single one with some parametrization of the test cases to test each of the 4 variants (XY, XYM, XYZ, XYZM) ?

Yes, indeed, there is a lot of copy/paste. I also thought about trying to come up with something to parameterize the tests, but I didn't go through with it. Why? I noticed that some tests were not always readable when you add several layers, sometimes you don't know on which layers you really snap on, etc.
I really preferred to isolate each tests to have a readability even for non developers.
I've done that for other tests lately, but if you think I should do that again, I will.

rouault commented:
> I noticed that some tests were not always readable when you add several layers, sometimes you don't know on which layers you really snap on, etc.

I guess you could have a generic class TestQgsMapToolCircleAbstract with some parametrization of the layer name, default values and expected result, and lightweight deriving classes TestQgsMapToolCircleXY, TestQgsMapToolCircleAbstractXYZ, TestQgsMapToolCircleAbstractXYM, TestQgsMapToolCircleAbstractXYZM

lbartoletti committed:
rewrite CircleM  (#15)

This commit rewrite the tests for QgsCircle mapTool.

Instead of having 4 separate files, we merged in one using data driven testing thanks to QTest. The goal is to reduce the number of lines, while remaining understandable for non-developers if necessary.

Co-authored-by: Antoine Facchini <antoine.facchini@oslandia.com>

lbartoletti committed:
fix cmake

lbartoletti committed:
remove useless file

lbartoletti committed:
prevent crash for map in a thread tests

lbartoletti committed:
reinstate paul copyright git add tests/src/app/testqgsmaptoolcircle.cpp

lbartoletti committed:
fix mvalue

lbartoletti committed:
minor changes in test

rouault commented on src/app/qgsmaptooladdcircle.cpp:
```diff
@@ -50,6 +50,13 @@ void QgsMapToolAddCircle::deactivate()
       break;
     }
+    if ( QgsWkbTypes::hasM( point.wkbType() ) &&
+         point.z() != defaultMValue() )
```
shouldn't that be ?
```suggestion
         point.m() != defaultMValue() )
```
not caught by tests ?

lbartoletti replied:
Right!
Yes it is not tested for now (as for the Z). It will be tested soon after my others PR on M/Z editing.

lbartoletti replied:
Mmm, I should split up my tests and introduce the snapping tests for circle into this PR

lbartoletti committed:
fix typos

lbartoletti committed:
this code is not needed since z/m are captured from QgsMapToolCapture::mapPoint and z/m are converted by QgsCircle

lbartoletti committed:
also return m value in QgsMapToolCapture::mapPoint

lbartoletti committed:
revert commit on qgsmaptoolcapture.cpp file

rouault commented:
Is this PR good to merge ?

lbartoletti commented:
> Is this PR good to merge ?

Yes, please.

lbartoletti commented:
Thanks @rouault
